File: HighlyDivisibleTriangularNumber.py
#!/usr/bin/python3
import math

primeList = [2]


def main():
    for i in primeGenerator():
        if(i > 10000):
            break
        primeList.append(i)

    i = 0
    counter = 1
    t = 1
    while counter <= 500:
        counter = 1
        i += 1
        t = TriangularNumber(i)
        tt = t
        for p in primeList:
            if p**2 > tt:
                counter *= 2
                break
            
            exponent = 1
            while tt % p == 0:
                exponent += 1
                tt = tt / p
            else:
                if exponent > 1: counter *= exponent
                if tt == 1: break
    else:
        print(t)

def TriangularNumber(n):
    return (n * (n + 1)) // 2


def isPrime(n):
    if n == 1:
        return False
    if n < 4:
        return True
    if n % 2 == 0:
        return False
    if n < 9:
        return True
    if n % 3 == 0:
        return False
    f = 5
    r = math.floor(math.sqrt(n))
    while f <= r:
        if(n % f == 0):
            return False
        if(n % (f + 2) == 0):
            return False
        f += 6
    else:
        return True


def primeGenerator():
    counter = 3
    while True:
        if(isPrime(counter)):
            yield counter
        counter += 2


if __name__ == '__main__':
    main()

File: HighlyDivisibleTriangularNumber.cpp
// This file is written by AlirezaKamyab
// Answer to problem projecteuler.net/problem=12
#include <iostream>
#include <cmath>
using namespace std;

bool* primeList(int limit);
int divisorsCount(bool* primeList, int number);

int main(){
    int limit = 500; 
    bool* primes = primeList(1000);

    int n = 1;
    int n1 = n + 1;
    while (true)
    {
        int countN = divisorsCount(primes, (n % 2 == 0)? n / 2 : n); 
        int countN1 = divisorsCount(primes,  (n1 % 2 == 0)? n1 / 2 : n1);

        if(countN * countN1 >= limit){
            cout << n * (n1) / 2 << endl;
            return 0;
        }
        n++;
        n1 = n + 1;
    }
    
    return 0;
}

int divisorsCount(bool* primeList, int number){
    int count = 1;
    for(int i = 1; (2 * i + 1) * (2 * i + 1) <= number; i++){
        if(primeList[i] == true) continue;
        int n = number;
        int exponent = 0;
        while(n % (i * 2 + 1) == 0){
            exponent++;
            n /= (i * 2) + 1;
        }

        count *= (exponent + 1);
    }

    if(number % 2 == 0){
        int exponent = 0;
        while(number % 2 == 0){
            exponent++;
            number /= 2;
        }
        count *= (exponent + 1);
    }

    return count;
}

bool* primeList(int limit){
    int sieveBound = (limit - 1) / 2;
    int crossLimit = (floor(sqrt(limit)) - 1) / 2;
    bool* primes = new bool[sieveBound] { 0 };

    for(int i = 1; i <= crossLimit; i++){
        if(primes[i] == true) continue;
        for(int j = 2 * i * (i + 1); j <= sieveBound; j += (2 * i) + 1){
            primes[j] = true;
        }
    }

    return primes;
}